File: wc/src/vn/gmgroup/wc/FilepathHandler.java
/*
 * Copyright (c) Artnet Gmsolution. All rights reserved.
 *
 * 文字列のファイルパスからGlobパータンを考慮し、ファイルパスを抽出するクラス
 */
package vn.gmgroup.wc;

import java.io.File;
import java.nio.file.*;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class FilepathHandler {

    /**
     * 指定された入力文字列に対してファイルの検索を行い、検索結果を返す。
     * 入力文字列はglobパターンである場合、それに応じてファイルを抽出します。Globパターンでない場合
     * 入力文字列をリスト内の単一の要素として返します。
     *
     * @param rawFilePath Globパータンである可能性があるロールファイルパス。
     * @return Globパターンにマッチするファイルパス一覧です。Globパータンではない場合、rawFilePath自体
     */
    private static List<String> findFilesWithGlobPattern(String rawFilePath) {
        List<String> result = new ArrayList<>();

        try {
            // Globパターンであるかどうかをチェックする
            if (rawFilePath.contains("*") || rawFilePath.contains("?") || rawFilePath.contains("[") || rawFilePath.contains("{")) {
                PathMatcher matcher;
                String normalizedPattern = rawFilePath.replace("\\", "/");
                String basePart = normalizedPattern.substring(0, rawFilePath.lastIndexOf("/") > 0 ?
                        rawFilePath.lastIndexOf("/") : rawFilePath.length());

                Path path = Paths.get(basePart);
                Path basePath;
                String pattern;

                if (path.isAbsolute()) {
                    basePath = path;
                    pattern = "" + normalizedPattern.replaceFirst(basePart, "");
                } else {
                    basePath = Paths.get(".");
                    pattern = rawFilePath;
                }

                matcher = FileSystems.getDefault().getPathMatcher("glob:" + pattern);

                try (Stream<Path> stream = Files.walk(basePath)) {
                    List<String> matches = stream
                            .filter(Files::isRegularFile)
                            .map(basePath::relativize)
                            .filter(matcher::matches)
                            .map(foundPath -> basePath.resolve(foundPath).toString())
                            .collect(Collectors.toList());

                    if (!matches.isEmpty()) {
                        return matches;
                    }
                }
            }
            // Globパータンではない場合、ローファイルパスをそのまま返す
            result.add(rawFilePath);
        } catch (Exception ex) {
            result.add(rawFilePath);
        }
        return result;
    }

    public static List<String> findFiles(List<String> rawFilePaths) {
        List<String> result = new ArrayList<>();
        for (String rawFilePath : rawFilePaths) {
            result.addAll(findFilesWithGlobPattern(rawFilePath));
        }
        return result;
    }
}

File: cat/src/main/kotlin/vn/gmgroup/cat/FilepathHandler.kt
/*
 * Copyright (c) Artnet Gmsolution. All rights reserved.
 *
 * 文字列のファイルパスからGlobパータンを考慮し、ファイルパスを抽出するクラス
 */
package vn.gmgroup.cat

import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

class FilepathHandler {
    companion object {
        private fun searchFilesByGlobPattern(inputPattern: String): List<String> {
            val result: MutableList<String> = mutableListOf()

            if (inputPattern.any { it in listOf('*', '?', '[', ']') }) {

                val normalizedPattern = inputPattern.replace("\\", "/")
                val (dirPath, globPattern) = extractDirAndPattern(normalizedPattern)

                val matcher = FileSystems.getDefault().getPathMatcher("glob:$globPattern")

                val paths = Files.walk(dirPath)
                try {
                    paths
                        .filter { path ->
                            Files.isRegularFile(path) && matcher.matches(path.fileName)
                        }
                        .forEach { path ->
                            result.add(path.toAbsolutePath().toString())
                        }
                } finally {
                    paths.close()
                }
            } else {
                result.add(inputPattern)
            }

            return result
        }

        private fun extractDirAndPattern(inputPattern: String): Pair<Path, String> {
            val lastSlashIndex = inputPattern.lastIndexOf('/')
            return if (lastSlashIndex != -1) {
                val dirPart = inputPattern.substring(0, lastSlashIndex)
                val globPart = inputPattern.substring(lastSlashIndex + 1)
                Pair(Paths.get(dirPart).toAbsolutePath(), "$globPart")
            } else {
                Pair(Paths.get(".").toAbsolutePath(), "$inputPattern")
            }
        }

        fun searchFilesByMultipleGlobPatterns(patterns: Array<String>): Set<String> {
            val allFilePaths = mutableSetOf<String>()

            patterns.forEach { pattern ->
                val matchedFiles = searchFilesByGlobPattern(pattern)
                allFilePaths.addAll(matchedFiles)
            }

            return allFilePaths
        }
    }
}
